flake8_trio/visitors/visitors.py

```diff
@@ -0,0 +1,415 @@
+                )
+            ):
+                self.error(item.context_expr, var_name)
+
+    visit_AsyncWith = visit_With
+
+
+# used in 113
+def _get_identifier(node: ast.expr) -> str:
+    if isinstance(node, ast.Name):
+        return node.id
+    if isinstance(node, ast.Attribute):
+        return node.attr
+    return ""
+
+
+# used in 113 and 114
+STARTABLE_CALLS = (
+    "run_process",
+    "serve_ssl_over_tcp",
+    "serve_tcp",
+    "serve_listeners",
+    "serve",
+)
+
+
+@error_class
+class Visitor113(Flake8TrioVisitor):
+    error_codes = {
+        "TRIO113": (
+            "Dangerous `.start_soon()`, process might not run before `__aenter__` "
+            "exits. Consider replacing with `.start()`."
+        ),
+    }
+
+    def __init__(self, *args: Any, **kwargs: Any):
+        super().__init__(*args, **kwargs)
+        self.async_function = False
+        self.asynccontextmanager = False
+        self.aenter = False
+
+    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef):
+        self.save_state(node, "aenter")
+
+        self.aenter = node.name == "__aenter__" or any(
+            _get_identifier(d) == "asynccontextmanager" for d in node.decorator_list
+        )
+
+    def visit_Yield(self, node: ast.Yield):
+        self.aenter = False
+
+    def visit_Call(self, node: ast.Call) -> None:
+        def is_startable(n: ast.expr, *startable_list: str) -> bool:
+            if isinstance(n, ast.Name):
+                return n.id in startable_list
+            if isinstance(n, ast.Attribute):
+                return n.attr in startable_list
+            if isinstance(n, ast.Call):
+                return any(is_startable(nn, *startable_list) for nn in n.args)
+            return False
+
+        if (
+            self.aenter
+            and is_nursery_call(node.func, "start_soon")
+            and len(node.args) > 0
+            and is_startable(
+                node.args[0],
+                *STARTABLE_CALLS,
+                *self.options.startable_in_context_manager,
+            )
+        ):
+            self.error(node)
+
+
+# Checks that all async functions with a "task_status" parameter have a match in
+# --startable-in-context-manager. Will only match against the last part of the option
+# name, so may miss cases where functions are named the same in different modules/classes
+# and option names are specified including the module name.
+@error_class
+class Visitor114(Flake8TrioVisitor):
+    error_codes = {
+        "TRIO114": (
+            "Startable function {} not in --startable-in-context-manager parameter "
+            "list, please add it so TRIO113 can catch errors using it."
+        ),
+    }
+
+    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef):
+        if any(
+            isinstance(n, ast.arg) and n.arg == "task_status"
+            for n in self.walk(
+                *node.args.args, *node.args.posonlyargs, *node.args.kwonlyargs
```

Comment:
Pretty sure this *shouldn't* check posonlyargs actually, IIRC Trio always passes it by name.

Reply:
Yep, the doc confirms
> The async_fn must accept a `task_status` keyword argument, [...]

https://trio.readthedocs.io/en/stable/reference-core.html#trio.Nursery.start